consensus/ibft/ibft.go

```diff
@@ -187,28 +194,109 @@ type ConsensusMechanism interface {
+	ibft *Ibft
+
+	// hookMap is the collection of registered hooks
+	hookMap map[HookType]func(interface{}) error
+
+	// Used for easy lookups
+	mechanismType MechanismType
+
+	// Available periods
+	From uint64
+	To   *uint64
+}
+
+// initializeParams initializes mechanism parameters from chain config
+func (base *BaseConsensusMechanism) initializeParams(params *IBFTFork) error {
+	if params == nil {
+		return nil
+	}
+
+	base.From = params.From.Value
+
+	if params.To != nil {
+		if params.To.Value < base.From {
+			return fmt.Errorf(
+				`"to" must be grater than or equal to from: from=%d, to=%d`,
+				base.From,
+				params.To.Value,
+			)
+		}
+
+		base.To = &params.To.Value
+	}
+
+	return nil
+}
+
+// GetType implements the ConsensusMechanism interface method
+func (base *BaseConsensusMechanism) GetType() MechanismType {
+	return base.mechanismType
+}
+
+// GetHookMap implements the ConsensusMechanism interface method
+func (base *BaseConsensusMechanism) GetHookMap() map[HookType]func(interface{}) error {
+	return base.hookMap
+}
+
+// IsInRange returns indicates if the given blockNumber is between from and to
+func (base *BaseConsensusMechanism) IsInRange(blockNumber uint64) bool {
+	// not ready
+	if blockNumber < base.From {
+		return false
+	}
+
+	// expired
+	if base.To != nil && *base.To < blockNumber {
+		return false
+	}
+
+	return true
+}
+
+// IBFT Fork represents setting in params.engine.ibft of genesis.json
+type IBFTFork struct {
+	Type       MechanismType      `json:"type"`
+	Deployment *common.JSONNumber `json:"deployment,omitempty"`
+	From       common.JSONNumber  `json:"from"`
+	To         *common.JSONNumber `json:"to,omitempty"`
+}
+
 // ConsensusMechanismFactory is the factory function to create a consensus mechanism
-type ConsensusMechanismFactory func(ibft *Ibft) (ConsensusMechanism, error)
+type ConsensusMechanismFactory func(ibft *Ibft, params *IBFTFork) (ConsensusMechanism, error)
 
 var mechanismBackends = map[MechanismType]ConsensusMechanismFactory{
 	PoA: PoAFactory,
 	PoS: PoSFactory,
 }
 
 // runHook runs a specified hook if it is present in the hook map
-func (i *Ibft) runHook(hookName string, hookParams interface{}) error {
-	// Grab the hook map
-	hookMap := i.mechanism.GetHookMap()
+func (i *Ibft) runHook(hookName HookType, height uint64, hookParam interface{}) error {
+	for _, mechanism := range i.mechanisms {
+		if !mechanism.IsAvailable(hookName, height) {
+			continue
+		}
 
-	// Grab the actual hook if it's present
-	hook, ok := hookMap[hookName]
-	if !ok {
-		// hook not found, continue
-		return ErrMissingHook
+		// Grab the hook map
+		hookMap := mechanism.GetHookMap()
+
+		// Grab the actual hook if it's present
+		hook, ok := hookMap[hookName]
+		if !ok {
+			// hook not found, continue
+			continue
+		}
+
+		// Run the hook
+		if err := hook(hookParam); err != nil {
+			return fmt.Errorf("error occurres during a call of %s hook in %s: %w", hookName, mechanism.GetType(), err)
```

Comment:
Typo here `occurres`, should be `occurred`

Reply:
Fixed at 4dd5210951014804bcc12ecbb507d80374ba1121